src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +271,417 @@ create_distributed_table(PG_FUNCTION_ARGS)
+	relation_close(relation, AccessShareLock);
+
+	/*
+	 * Get name of the table before possibly replacing it in
+	 * citus_add_local_table_to_metadata.
+	 */
+	char *tableName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	RangeVar *rangeVar = makeRangeVar(schemaName, tableName, -1);
+
+	/*
+	 * Before taking locks, convert the table into a Citus local table and commit
+	 * to allow shard split to see the shard.
+	 */
+	SendAddLocalTableToMetadataCommandOutsideTransaction(relationId);
+
+	/*
+	 * Lock target relation with a shard update exclusive lock to
+	 * block DDL, but not writes.
+	 *
+	 * If there was a concurrent drop/rename, error out by setting missingOK = false.
+	 */
+	bool missingOK = false;
+	relationId = RangeVarGetRelid(rangeVar, ShareUpdateExclusiveLock, missingOK);
+
+	if (PartitionedTable(relationId))
+	{
+		/* also lock partitions */
+		LockPartitionRelations(relationId, ShareUpdateExclusiveLock);
+	}
+
+	List *shardList = LoadShardIntervalList(relationId);
+
+	/*
+	 * It's technically possible for the table to have been concurrently
+	 * distributed just after citus_add_local_table_to_metadata and just
+	 * before acquiring the lock, so double check.
+	 */
+	if (list_length(shardList) != 1 ||
+		!IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("table was concurrently modified")));
+	}
+
+	/*
+	 * The table currently has one shard, we will split that shard to match the
+	 * target distribution.
+	 */
+	ShardInterval *shardToSplit = (ShardInterval *) linitial(shardList);
+
+	PropagatePrerequisiteObjectsForDistributedTable(relationId);
+
+	/*
+	 * Make sure that existing reference tables have been replicated to all the nodes
+	 * such that we can create foreign keys and joins work immediately after creation.
+	 * We do this after applying all essential checks to error out early in case of
+	 * user error.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	Oid colocatedTableId = InvalidOid;
+	if (colocationId != INVALID_COLOCATION_ID)
+	{
+		colocatedTableId = ColocatedTableId(colocationId);
+	}
+
+	List *workersForPlacementList;
+	List *shardSplitPointsList;
+
+	if (colocatedTableId != InvalidOid)
+	{
+		List *colocatedShardList = LoadShardIntervalList(colocatedTableId);
+
+		/*
+		 * Match the shard ranges of an existing table.
+		 */
+		shardSplitPointsList = HashSplitPointsForShardList(colocatedShardList);
+
+		/*
+		 * Find the node IDs of the shard placements.
+		 */
+		workersForPlacementList = WorkerNodesForShardList(colocatedShardList);
+	}
+	else
+	{
+		/*
+		 * Generate a new set of #shardCount shards.
+		 */
+		shardSplitPointsList = HashSplitPointsForShardCount(shardCount);
+
+		/*
+		 * Place shards in a round-robin fashion across all data nodes.
+		 */
+		List *workerNodeList = DistributedTablePlacementNodeList(NoLock);
```

Comment:
We use NoLock here, but I'm kind of thinking we should make sure that no more worker nodes get concurrently added after EnsureReferenceTablesExistOnAllNodes. Otherwise we do have an empty worker node and the remainder of the operation will fail. Maybe this call DistributedTablePlacementNodeList should be above EnsureReferenceTablesExistOnAllNodes with a RowShare lock. That way, we have some guarantee that all nodes have reference tables.